File: go/tasks.go
//go:build !debug && !tests

package main

import (
	"C"
	"fmt"

	"github.com/uhppoted/uhppote-core/uhppote"
)

func addTask(uu uhppote.IUHPPOTE, deviceID uint32, task *C.struct_Task) error {
	if task == nil {
		return fmt.Errorf("invalid argument (task) - expected valid pointer")
	}

	t, err := makeTask(task)
	if err != nil {
		return err
	} else if t == nil {
		return fmt.Errorf("invalid task (%v)", t)
	}

	if ok, err := uu.AddTask(deviceID, *t); err != nil {
		return err
	} else if !ok {
		return fmt.Errorf("%v: add-task failed for %v", deviceID, t.Task)
	}

	return nil
}

func refreshTaskList(uu uhppote.IUHPPOTE, deviceID uint32) error {
	ok, err := uu.RefreshTaskList(deviceID)
	if err != nil {
		return err
	} else if !ok {
		return fmt.Errorf("%v: refresh-tasklist failed", deviceID)
	}

	return nil
}

func clearTaskList(uu uhppote.IUHPPOTE, deviceID uint32) error {
	ok, err := uu.ClearTaskList(deviceID)
	if err != nil {
		return err
	} else if !ok {
		return fmt.Errorf("%v: clear-tasklist failed", deviceID)
	}

	return nil
}

File: tests/c++/tasks.cpp
#include <iomanip>
#include <iostream>

#include "../include/uhppoted.hpp"
#include "tests.hpp"

using namespace std;

bool addTask(uhppoted &u) {
    task task = {
        .task = 4,
        .door = 3,
        .from = "2022-02-01",
        .to = "2022-06-30",
        .monday = true,
        .tuesday = false,
        .wednesday = true,
        .thursday = true,
        .friday = false,
        .saturday = false,
        .sunday = true,
        .at = "09:45",
        .cards = 11,
    };

    u.add_task(DEVICE_ID, task);

    vector<result> rs = {};

    return evaluate("add-task", rs);
}

bool refreshTaskList(uhppoted &u) {
    u.refresh_tasklist(DEVICE_ID);

    vector<result> rs = {};

    return evaluate("refresh-tasklist", rs);
}

bool clearTaskList(uhppoted &u) {
    u.clear_tasklist(DEVICE_ID);

    vector<result> rs = {};

    return evaluate("clear-tasklist", rs);
}
